tree/30_generic_methods/command_execution_once.cf

```diff
@@ -0,0 +1,96 @@
+#
+# @parameter command        The command to run
+# @parameter ok_codes       List of codes that produce a kept status separated with commas (ex: 1,2,5). Defaults to 0.
+# @parameter until          Try to execute the command until a particular state: 'ok', 'any' (defaults to 'any')
+# @parameter_constraint until "allow_empty_string" : true
+# @parameter_constraint until "select" : [ "", "any", "ok" ]
+# @parameter unique_id      To identify the command (allow to change the command without losing track)
+# 
+# @class_prefix command_execution_once
+# @class_parameter command
+
+bundle agent command_execution_once(command, ok_codes, until, unique_id)
+{
+  vars:
+      "c_unique_id"       string => canonify("${unique_id}");
+      "old_class_prefix"  string => canonify("command_execution_once_${command}");
+      "args"               slist => { "${command}", "${ok_codes}", "${until}", "${unique_id}" };
+      "report_param"      string => join("_", args);
+      "full_class_prefix" string => canonify("command_execution_once_${report_param}");
+      "class_prefix"      string => string_head("${full_class_prefix}", "1000");
+
+      "ok_list"            slist => splitstring("${ok_codes}", "\s*,\s*", "256");
+      "repaired_list"      slist => {};
+
+  defaults:
+      "until"    string => "any", if_match_regex => "";
+      "ok_codes" string => "0", if_match_regex => "";
+
+  classes:
+      "until_ok"    expression => strcmp("${until}", "ok");
+
+      "reserved"    expression => strcmp("${reserve[1]}", "ok");
+      "already_set" expression => strcmp("${reserve[1]}", "failed");
+      "set_ok"      expression => strcmp("${set[1]}", "ok");
+
+      "${c_condition}" expression => "set_ok",
+                            scope => "namespace";
+
+    any::
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+  methods:
+      "reserve the flag" usebundle => reserve_flag("command_execution_once_${c_unique_id}"),
+                         useresult => "reserve";
+      "set the flag"     usebundle => set_flag("command_execution_once_${c_unique_id}"),
+                         useresult => "set",
+                                if => "reserved.${class_prefix}_repaired";
+
+    pass3.reserved.!set_ok::
+      "failure" usebundle => _classes_failure("${old_class_prefix}");
+      "failure" usebundle => _classes_failure("${class_prefix}");
+
+    pass3.already_set::
+      "success" usebundle => _classes_success("${old_class_prefix}");
+      "success" usebundle => _classes_success("${class_prefix}");
+
+    pass3::
+      "clean"   usebundle => _classes_sanitize("${old_class_prefix}");
+      "clean"   usebundle => _classes_sanitize("${class_prefix}");
+
+      "report"
+        usebundle  => _log_v3("Execute the command '${command}' once", "${command}", "${old_class_prefix}", "${class_prefix}", @{args});
+
+  commands:
+    reserved::
+      "${command}"
```

Comment:
why not using command_execution ?

Reply:
Makes sense indeed (with `command_execution_result`)